cmd/joy/release_promote_filters_test.go

```diff
@@ -0,0 +1,222 @@
+			},
+		},
+		{
+			Name: "all flag with keep pre release",
+			Args: []string{"--all", "--keep-pre-release"},
+			Releases: []TestCrossRelease{
+				{
+					Name:   "pre-release-target",
+					Source: &TestFile{Path: "alpha-source.yaml", Content: "{spec: {version: 1.2.3}}"},
+					Target: &TestFile{Path: "alpha-target.yaml", Content: "{spec: {version: 1.0.0-some-feature}}"},
+				},
+				{
+					Name:   "nil-target",
+					Source: &TestFile{Path: "beta-source.yaml", Content: "{spec: {version: 3.2.1}}"},
+					Target: nil,
+				},
+			},
+			Expectations: func(t *testing.T, files []*yml.File) {
+				require.Len(t, files, 1)
+
+				dir, filename := filepath.Split(files[0].Path)
+				_, targetDir := filepath.Split(filepath.Clean(dir))
+
+				require.Equal(t, "target/beta-source.yaml", filepath.Join(targetDir, filename))
+				require.Equal(t, "{spec: {version: 3.2.1}}\n", string(files[0].Yaml))
+			},
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.Name, func(t *testing.T) {
+			makeFile := func(t *testing.T, name, content string) *yml.File {
+				t.Helper()
+
+				f, err := yml.NewFile(name, []byte(content))
+				require.NoError(t, err)
+
+				return f
+			}
+
+			v1alpha1Release := func(release TestCrossRelease, env *v1alpha1.Environment, file *TestFile) *v1alpha1.Release {
+				if file == nil {
+					return nil
+				}
+				return &v1alpha1.Release{
+					ReleaseMetadata: v1alpha1.ReleaseMetadata{Name: release.Name},
+					Environment:     env,
+					File:            makeFile(t, filepath.Join(env.Dir, file.Path), file.Content),
+					Project:         &v1alpha1.Project{},
+				}
+			}
+
+			// We are giving a valid local path so that the relative path functionality
+			// continues to work for promotions to nil targets
+			cwd, err := os.Getwd()
+			require.NoError(t, err)
+
+			source := &v1alpha1.Environment{
+				EnvironmentMetadata: v1alpha1.EnvironmentMetadata{Name: "source"},
+				Dir:                 filepath.Join(cwd, "source"),
+			}
+			target := &v1alpha1.Environment{
+				EnvironmentMetadata: v1alpha1.EnvironmentMetadata{Name: "target"},
+				Spec: v1alpha1.EnvironmentSpec{
+					Promotion: v1alpha1.Promotion{
+						FromEnvironments: []string{source.Name},
+					},
+				},
+				Dir: filepath.Join(cwd, "target"),
+			}
+
+			envs := []*v1alpha1.Environment{source, target}
+
+			cat := &catalog.Catalog{
+				Environments: envs,
+				Releases: cross.ReleaseList{
+					Environments: envs,
+				},
+			}
+
+			for _, release := range tc.Releases {
+				cat.Releases.Items = append(cat.Releases.Items, &cross.Release{
+					Name: release.Name,
+					Releases: []*v1alpha1.Release{
+						v1alpha1Release(release, source, release.Source),
+						v1alpha1Release(release, target, release.Target),
+					},
+				})
+			}
+
+			for _, cross := range cat.Releases.Items {
+				for _, rel := range cross.Releases {
+					if rel == nil {
+						continue
+					}
+					require.NoError(t, yaml.Unmarshal(rel.File.Yaml, &rel))
+				}
+			}
```

Comment:
Wouldn't it be possible to merge those two loops?

Reply:
It is totally possible but it hurts readability in my opinion

Reply:
If there's anything, current implementation left me more confused, not understanding why looping twice over the same items.  Here's what it would look like, not finding it less readable really:
```go
			for _, release := range tc.Releases {
				crossRel := cross.Release{
					Name: release.Name,
					Releases: []*v1alpha1.Release{
						v1alpha1Release(release, source, release.Source),
						v1alpha1Release(release, target, release.Target),
					},
				}
				cat.Releases.Items = append(cat.Releases.Items, &crossRel)
				for _, rel := range crossRel.Releases {
					if rel == nil {
						continue
					}
					require.NoError(t, yaml.Unmarshal(rel.File.Yaml, &rel))
				}
			}
```

Reply:
I guess it is a matter of opinion. I will change it!